### PR TITLE
Fix Ascend NPU RMSNorm and fused-attention mask shapes

### LINKED ISSUE #14380: [Bug] Ascend NPU: RMSNorm crashes with elementwise_affine=False; _native_npu FA rejects [B, N, 1, Skv] masks (LTX-2)
### Describe the bug

When running LTX-2 with `attn_backend=_native_npu` on Ascend NPU, two incompatibilities show up:

1. **RMSNorm** – layers with `elementwise_affine=False` leave `weight=None`, but `torch_npu.npu_rms_norm` requires a gamma tensor → crash (`gamma is None`).

2. **Fused attention mask** – LTX cross-attn uses masks shaped `[B, N, 1, Skv]` (e.g. `[1, 32, 1, 1024]`). Ascend FA does not broadcast the singleton query-length dim the way SDPA does, so `_maybe_modify_attn_mask_npu` must expand it to `[B, N, Sq, Skv]`. Today only `[B, 1, 1, Skv]` is expanded.

These bugs were discovered while running LTX-2.3 text-to-audio-video FlowGRPO training on Ascend NPU using [verl-omni](https://github.com/verl-project/verl-omni/blob/main/examples/flowgrpo_trainer/ltx2/run_ltx2_3_t2av_lora_npu.sh).

### Reproduction

Requires Ascend NPU + `torch_npu`. Minimal sketches of both failure modes:

```python
import torch
from diffusers.models.normalization import RMSNorm
from diffusers.models.attention_dispatch import (
    AttentionBackendName,
    attention_backend,
    dispatch_attention_fn,
)

# --- Bug 1: RMSNorm with elementwise_affine=False ---
norm = RMSNorm(dim=64, eps=1e-6, elementwise_affine=False).to("npu")
x = torch.randn(2, 16, 64, device="npu", dtype=torch.float16)
# Crashes: npu_rms_norm called with weight=None
y = norm(x)

# --- Bug 2: FA mask [B, N, 1, Skv] under _native_npu ---
B, Sq, Skv, N, D = 1, 384, 1024, 32, 64
q = torch.randn(B, Sq, N, D, device="npu", dtype=torch.float16)
k = torch.randn(B, Skv, N, D, device="npu", dtype=torch.float16)
v = torch.randn(B, Skv, N, D, device="npu", dtype=torch.float16)
attn_mask = torch.zeros(B, N, 1, Skv, device="npu", dtype=torch.float16)  # LTX-style

with attention_backend(AttentionBackendName._NATIVE_NPU):
    # Fails: Ascend FA expects Sq on dim=-2, not a singleton 1
    out = dispatch_attention_fn(q, k, v, attn_mask=attn_mask)
```

### Logs

# Error 1

Image

# Error 2

Image

### System Info

- OS: Linux aarch64
- Hardware: Ascend NPU
- Python: 3.11
- torch: 2.10.0
- torch_npu: 2.10.0
- CANN: 9.0.0
- diffusers: main

### Who can help?

@yiyixuxu @sayakpaul

### PR DESCRIPTION
**Pull request by mengchengTang**
# What does this PR do?

Fixes two Ascend NPU incompatibilities hit when running LTX-2 with `attn_backend=_native_npu`:

1. **RMSNorm** with `elementwise_affine=False` leaves `weight=None`. `torch_npu.npu_rms_norm` requires a gamma tensor, so we only use the fused kernel when `weight is not None`, and otherwise fall back to the existing PyTorch path (same as CUDA). This is a silent perf change for that config; the previous path would crash, so it is unlikely anyone relied on it. The fallback still ends with `hidden_states.to(input_dtype)`.

2. **Fused attention mask**: Ascend FA does not broadcast a singleton query-length dim. Expand masks shaped `[B, 1|N, 1, Skv]` (e.g. LTX cross-attn `[B, N, 1, Skv]`) to `[B, 1|N, Sq, Skv]`, generalizing the existing `[B, 1, 1, Skv]` handling. The head dim is restricted to `1` or `N` (`query.shape[2]`) so unrelated head counts are not expanded.

## Screenshots

**Error 1 — RMSNorm / `gamma` is None**

image

**Error 2 — FA mask shape `[1, 32, 1, 1024]`**

image

Fixes #14380 

### Environment

| Item | Version / Configuration |
| --- | --- |
| OS | Linux aarch64 |
| Hardware | Ascend NPU |
| Python | 3.11 |
| PyTorch | 2.10.0 |
| torch-npu | 2.10.0 |
| CANN | 9.0.0 |
| diffusers | `main` |

### Test results

| Scope | Test | Before the fix | After the fix | Result |
| --- | --- | --- | --- | --- |
| End-to-end | [`dg845/LTX-2.3-Diffusers` with verl-omni LTX-2.3 FlowGRPO](https://github.com/verl-project/verl-omni/blob/main/examples/flowgrpo_trainer/ltx2/run_ltx2_3_t2av_lora_npu.sh) | Rollout or training stops at the RMSNorm or fused-attention mask error | Rollout and training proceed through both affected paths without error | ✓ Passed |
| Minimal reproduction | `RMSNorm(elementwise_affine=False)` from #14380 | `npu_rms_norm` receives `weight=None` and crashes | Falls back to the PyTorch RMSNorm path and completes successfully | ✓ Passed |
| Minimal reproduction | Native NPU attention with mask `[B, N, 1, Skv]` from #14380 | Ascend fused attention rejects the singleton query dimension | Mask is expanded to `[B, N, Sq, Skv]` and attention completes successfully | ✓ Passed |

## Before submitting
- [x] Did you use an AI agent (Claude Code, Codex, Cursor, etc.) to help with this PR? If so:
  - [x] Did you read the [Coding with AI agents](https://huggingface.co/docs/diffusers/main/en/conceptual/contribution#coding-with-ai-agents) guide?
  - [x] Did you run the [`self-review`](https://github.com/huggingface/diffusers/blob/main/.ai/skills/self-review/SKILL.md) skill on the diff?
  - [x] Did you share the final self-review notes in the PR description or a comment?
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/diffusers/main/en/conceptual/contribution)?
- [x] Did you read our [philosophy doc](https://huggingface.co/docs/diffusers/main/en/conceptual/philosophy)? (important for complex PRs)
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?
- [ ] Are you the author (or part of the team) of the model/pipeline (only applicable for model/pipeline related PRs)?

## Who can review?

@yiyixuxu @sayakpaul — Ascend NPU / `_native_npu` attention + RMSNorm

### CODE REVIEW

**Review by ErenAta16 — COMMENTED** [review]
Ran both conditions over a spread of mask shapes. The `[B, N, 1, Skv]` case the PR is after does get picked up, and the added `shape[-1] == key.shape[1]` check is a genuine tightening — but two rows behave in ways the new comment doesn't cover.

`query` is `[B, Sq, N, D]` so `query.shape[1] == Sq`; `B=2, N=4, Sq=7, Skv=11`:

| mask shape | before | after | result shape (after) |
|---|---|---|---|
| `[B, 1, 1, Skv]` | expand | expand | `(2, 1, 7, 11)` |
| `[B, N, 1, Skv]` — the target | skip | **expand** | `(2, 4, 7, 11)` |
| `[B, 1, 1, Skv+3]` | **expand** | skip | `(2, 1, 1, 14)` |
| `[B, N, 1, Skv+3]` | skip | skip | `(2, 4, 1, 14)` |
| `[B, 99, 1, Skv]` — nonsense head count | skip | **expand** | `(2, 99, 7, 11)` |
| `[B, 1, Sq, Skv]` — already correct | skip | skip | unchanged |
| `[B+5, 1, 1, Skv]` — wrong batch | expand | expand | `(7, 1, 7, 11)` |

Row 3 is a behaviour change worth stating: a 4-D mask whose last dim doesn't match `key.shape[1]` used to be expanded anyway and now falls through to `npu_fusion_attention` as-is. That's more correct — silently expanding a mismatched mask was the worse option — but it moves where the failure surfaces from this function to the kernel.

Row 5 is the one I'd tighten. `expand(-1, -1, Sq, -1)` accepts any `shape[1]`, so dropping `shape[1] == 1` without replacing it means a mask with a head count that matches nothing gets reshaped and handed to the kernel. `shape[1] in (1, query.shape[2])` would keep the intent ("1 or N") and reject the rest here rather than downstream.

Row 7 is pre-existing on both sides — the 2-D branch checks `attn_mask.shape[0] == query.shape[0]` and the 4-D branch never has. Not this PR's problem, just noting the asymmetry since the 4-D branch is being touched.

On the `normalization.py` half: the fix looks right, and `self.bias` can only be non-`None` when `elementwise_affine=True` (it's created inside that branch), so the `if self.bias is not None` left in the NPU path is consistent with `self.weight is not None` now gating it. Worth saying out loud in the PR body that NPU users with `elementwise_affine=False` now take the pure-PyTorch path rather than the fused kernel — that's the correct trade, but it's a silent perf change for anyone who was previously… well, crashing, so probably nobody. The `else` branch ends with `hidden_states.to(input_dtype)`, so the dtype contract holds on the fallback.

Shape table produced with plain torch tensors on CPU (`_maybe_modify_attn_mask_npu`'s reshaping is device-independent). I have no Ascend hardware, so I can't confirm what `npu_fusion_attention` does with the row-3 and row-5 masks it now receives — that's the part worth a check from someone with an NPU.

**Comment by mengchengTang** [comment]
> Ran both conditions over a spread of mask shapes. The `[B, N, 1, Skv]` case the PR is after does get picked up, and the added `shape[-1] == key.shape[1]` check is a genuine tightening — but two rows behave in ways the new comment doesn't cover.
> 
> `query` is `[B, Sq, N, D]` so `query.shape[1] == Sq`; `B=2, N=4, Sq=7, Skv=11`:
> 
> mask shape	before	after	result shape (after)
> `[B, 1, 1, Skv]`	expand	expand	`(2, 1, 7, 11)`
> `[B, N, 1, Skv]` — the target	skip	**expand**	`(2, 4, 7, 11)`
> `[B, 1, 1, Skv+3]`	**expand**	skip	`(2, 1, 1, 14)`
> `[B, N, 1, Skv+3]`	skip	skip	`(2, 4, 1, 14)`
> `[B, 99, 1, Skv]` — nonsense head count	skip	**expand**	`(2, 99, 7, 11)`
> `[B, 1, Sq, Skv]` — already correct	skip	skip	unchanged
> `[B+5, 1, 1, Skv]` — wrong batch	expand	expand	`(7, 1, 7, 11)`
> Row 3 is a behaviour change worth stating: a 4-D mask whose last dim doesn't match `key.shape[1]` used to be expanded anyway and now falls through to `npu_fusion_attention` as-is. That's more correct — silently expanding a mismatched mask was the worse option — but it moves where the failure surfaces from this function to the kernel.
> 
> Row 5 is the one I'd tighten. `expand(-1, -1, Sq, -1)` accepts any `shape[1]`, so dropping `shape[1] == 1` without replacing it means a mask with a head count that matches nothing gets reshaped and handed to the kernel. `shape[1] in (1, query.shape[2])` would keep the intent ("1 or N") and reject the rest here rather than downstream.
> 
> Row 7 is pre-existing on both sides — the 2-D branch checks `attn_mask.shape[0] == query.shape[0]` and the 4-D branch never has. Not this PR's problem, just noting the asymmetry since the 4-D branch is being touched.
> 
> On the `normalization.py` half: the fix looks right, and `self.bias` can only be non-`None` when `elementwise_affine=True` (it's created inside that branch), so the `if self.bias is not None` left in the NPU path is consistent with `self.weight is not None` now gating it. Worth saying out loud in the PR body that NPU users with `elementwise_affine=False` now take the pure-PyTorch path rather than the fused kernel — that's the correct trade, but it's a silent perf change for anyone who was previously… well, crashing, so probably nobody. The `else` branch ends with `hidden_states.to(input_dtype)`, so the dtype contract holds on the fallback.
> 
> Shape table produced with plain torch tensors on CPU (`_maybe_modify_attn_mask_npu`'s reshaping is device-independent). I have no Ascend hardware, so I can't confirm what `npu_fusion_attention` does with the row-3 and row-5 masks it now receives — that's the part worth a check from someone with an NPU.

Thanks for the careful shape table — very helpful.

Row 5: Agreed. I’ve tightened the 4-D branch to require attn_mask.shape[1] in (1, query.shape[2]), so we only expand head dim 1 (broadcast) or N, and reject unrelated head counts here instead of handing them to the kernel.

Row 3 / shape[-1] == key.shape[1]: I’ve dropped that check to avoid the behaviour change you flagged (mismatched last-dim masks falling through instead of being expanded). Expanding a wrong-Skv mask is still imperfect, but keeping the previous expand behaviour reduces surprise vs. silently moving the failure into npu_fusion_attention. Happy to revisit if maintainers prefer the stricter guard.

Row 7: Acknowledged — leaving the pre-existing 2-D vs 4-D batch-check asymmetry alone in this PR.

normalization.py: Thanks for confirming. I’ve updated the PR description to note that with elementwise_affine=False we take the pure-PyTorch path instead of the fused kernel, and that the fallback still restores dtype via hidden_states.to(input_dtype).

I’ve also verified the updated path on Ascend NPU (including the LTX-2 _native_npu case this PR targets).

**Comment by mengchengTang** [comment]
@yiyixuxu @sayakpaul @zhtmike These two Ascend NPU fixes (#14288 and #14396) are issues I ran into while adapting verl-omni — could you take a look when you get a chance? Thanks!

**Comment by zhtmike** [comment]
Hi @mengchengTang , I have replied in the issues

**Review by DN6 — APPROVED** [review]
